File: Ashutosh-Singh/SlidingWindow/Maximum-Number-of-Vowels-in-a-Substring-of-Given-Length.cpp
#include <iostream>
#include <string>
#include <climits>
using namespace std;

class Solution {
public:
    bool isVowel(char c) {
        return (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u');
    }

    int maxVowels(string s, int k) {
        int n = s.size();
        int vowelCnt = 0;
        int maxCnt = INT_MIN;

        // Count vowels in the first window
        for (int i = 0; i < k; i++) {
            if (isVowel(s[i])) {
                vowelCnt++;
            }
        }
        maxCnt = max(maxCnt, vowelCnt);

        int left = 0;
        int right = k;

        // Slide the window across the string
        while (right < n) {
            if (isVowel(s[right])) {
                vowelCnt++;
            }
            if (isVowel(s[left])) {
                vowelCnt--;
            }

            maxCnt = max(maxCnt, vowelCnt);
            left++;
            right++;
        }

        return maxCnt;
    }
};

int main() {
    Solution solution;
    string s;
    int k;

    cout << "Enter the string: ";
    cin >> s;

    cout << "Enter the size of the window (k): ";
    cin >> k;

    int result = solution.maxVowels(s, k);

    cout << "The maximum number of vowels in any substring of length " << k << " is: " << result << endl;

    return 0;
}

File: Shantanu_Mamgain/Sliding-Window-Leetcode/Maximum Number of Vowels in a Substring of Given Length.java
class Solution {
    public int maxVowels(String s, int k) {
        int start = 0 ;
        int end = k-1;
        int maxLen = 0;
        int count = 0;
        for(int i = start; i<=end; i++){
            if(s.charAt(i)=='a'||s.charAt(i)=='e'||s.charAt(i)=='i'||s.charAt(i)=='o'||s.charAt(i)=='u'){
                count ++;
            }
        }
        maxLen = Math.max(maxLen, count);

        while(end < s.length()-1){
            end++;
            start++;
            if(s.charAt(start-1)=='a'||s.charAt(start-1)=='e'||s.charAt(start-1)=='i'||s.charAt(start-1)=='o'||s.charAt(start-1)=='u'){
                count --;
            }
            if(s.charAt(end)=='a'||s.charAt(end)=='e'||s.charAt(end)=='i'||s.charAt(end)=='o'||s.charAt(end)=='u'){
                count ++;
            }
            
            maxLen = Math.max(maxLen, count);
            
        }

        return maxLen;
    }

}
